rasa/nlu/featurizers/featurizer.py

```diff
@@ -1,48 +1,141 @@
+
+    def __init__(self, name: Text, config: Dict[Text, Any]) -> None:
+        """Instantiates a new featurizer.
+
+        Args:
+          config: configuration
+          name: a name that can be used as identifier, in case the configuration does
+            not specify an `alias`
+        """
+        super().__init__()
+        self.validate_config(config)
+        self._config = {**self.get_default_config(), **config}
+        self._identifier = self._config.get(FEATURIZER_CLASS_ALIAS, name)
+
+    @property
+    def identifier(self) -> Text:
+        """Returns the name of this featurizer.
+
+        Every feature created by this featurizer will contain this identifier as
+        `origin` information.
+        """
+        return self._identifier
+
+    @classmethod
+    @abstractmethod
+    def validate_config(cls, config: Dict[Text, Any]) -> None:
+        """Validates that the component is configured properly."""
+        ...
+
+    @classmethod
+    @abstractmethod
+    def validate_compatibility_with_tokenizer(cls, tokenizer: Tokenizer) -> None:
+        """Validate that the featurizer is compatible with the given tokenizer."""
+        # TODO: add (something like) this to recipe validation
+        # TODO: replace tokenizer by config of tokenizer to enable static check
+        ...
+
+    @abstractmethod
+    def process(self, messages: List[Message]) -> List[Message]:
+        """Featurizes all given messages in-place.
 
+        Args:
+          messages: messages to be featurized
+        Returns:
+          the same list with the same messages after featurization
+        """
+        ...
 
-class SparseFeaturizer(Featurizer):
-    pass
+    def process_training_data(self, training_data: TrainingData) -> TrainingData:
+        """Processes the training examples in the given training data in-place.
+
+        Args:
+          training_data: the training data
+        Returns:
+          same training data after processing
+        """
+        self.process(training_data.training_examples)
+        return training_data
+
+    def add_features_to_message(
+        self,
+        sequence: FeatureType,
+        sentence: Optional[FeatureType],
+        attribute: Text,
+        message: Message,
+    ) -> None:
+        """Adds sequence and sentence features for the attribute to the given message.
+
+        Args:
+          sequence: sequence feature matrix
+          sentence: sentence feature matrix
+          attribute: the attribute which both features describe
+          message: the message to which we want to add those features
+        """
+        for type, features in [
+            (FEATURE_TYPE_SEQUENCE, sequence),
+            (FEATURE_TYPE_SENTENCE, sentence),
+        ]:
+            if features is not None:
+                wrapped_feature = Features(features, type, attribute, self.identifier,)
+                message.add_features(wrapped_feature)
+
+    @staticmethod
+    def validate_configs_compatible(
```

Comment:
nit: should we move this up so it's bundled with the rest of the `validate_` methods?

Reply:
put that below the rest on purpose because it's not validating a single featurizer config but *only* whether they're compatible with each other